### PR TITLE
Enable hostpolicy on VM mode for -enableKubeArmorHostPolicy flag

### PR DESCRIPTION
**Pull request by seswarrajan**
Signed-off-by: Eswar Rajan Subramanian <eswar@accuknox.com>

### CODE REVIEW

File: KubeArmor/common/common.go [hunk 1]
```diff
@@ -322,7 +322,7 @@ func GetExternalIPAddr() string {
 // IsK8sLocal Function
 func IsK8sLocal() bool {
 	k8sConfig := os.Getenv("KUBECONFIG")
-	if _, err := os.Stat(filepath.Clean(k8sConfig)); err == nil {
+	if _, err := os.Stat(filepath.Clean(k8sConfig)); k8sConfig != "" && err == nil {
 		return true
 	}
```
Comment (nyrahul) [inline KubeArmor/common/common.go]: No point in doing a `os.Stat` and then checking if `k8sConfig != ""`!
```suggestion
        if k8sConfig != "" {
	    if _, err := os.Stat(filepath.Clean(k8sConfig)); err == nil {
	   	return true
	    }
	}
```

File: KubeArmor/core/kubeArmor.go [hunk 2]
```diff
@@ -321,7 +321,8 @@ func KubeArmor() {
 	// create a daemon
 	dm := NewKubeArmorDaemon()
 
-	if cfg.GlobalCfg.KVMAgent {
+	// Enable KubeArmorHostPolicy for both VM and KVMAgent
+	if !kl.IsK8sEnv() && (cfg.GlobalCfg.KVMAgent || cfg.GlobalCfg.HostPolicy) {
```
Comment (nyrahul) [inline KubeArmor/core/kubeArmor.go]: What happens in cases where user has `~/.kube/config` file, but still wants to use kubearmor in pure VM/systemd mode?
Just the presence of kubeconfig file will cause kubearmor to think that it is using k8s mode of operation!

**Review by nyrahul — CHANGES_REQUESTED** [review]

File: KubeArmor/core/kubeArmor.go (hunk 2 above)
Comment (seswarrajan) [reply KubeArmor/core/kubeArmor.go]: Yes. Possible. In that case it will deny/ignore to run as systemd mode.

Comment (nam-jaehyun) [reply KubeArmor/core/kubeArmor.go]: How can we know if a user wants to run KubeArmor as a systemd mode in such as case?
@seswarrajan What happens if a user uses -enableKubeArmorVm without kvm-service?

Comment (seswarrajan) [reply KubeArmor/core/kubeArmor.go]: @nam-jaehyun User can enable KubeArmor systemd mode only by -enableKubeArmorHostPolicy flag. It is not possible to run kubeArmor in VM with -enableKubeArmorVm flag as this flag tries to connect to kvm-service and if the connection fails, KubeArmor will terminate.

Comment (seswarrajan) [reply KubeArmor/core/kubeArmor.go]: @nyrahul In that case we need to ignore the IsK8sEnv() validation check in the code. But internally in the code while applying for new policies IsK8sEnv() check is still run.

**Review by nam-jaehyun — APPROVED** [review]
I'm okay with this PR.

File: KubeArmor/core/kubeArmor.go (hunk 2 above)
Comment (nyrahul) [reply KubeArmor/core/kubeArmor.go]: This has to be fixed holistically ... currently the code just allows to bypass the first hurdle. If the IsK8sEnv() check is present elsewhere, then VM/systemd mode cannot be used.

Can you check if kubearmor works in systemd mode (with policy enforcement) when there is ~/.kube/config in the path?

**Review by nyrahul — CHANGES_REQUESTED** [review]

File: KubeArmor/common/common.go [hunk 3]
```diff
@@ -23,6 +23,8 @@ import (
 // == Common == //
 // ============ //
 
+var IsNonK8sEnv bool
```
Comment (nyrahul) [inline KubeArmor/common/common.go]: Please fix the lint warnings.

**Review by nyrahul — CHANGES_REQUESTED** [review]

File: KubeArmor/Makefile [hunk 4]
```diff
@@ -39,7 +39,7 @@ run: build
 run-host-only: build
 	cd $(CRDDIR); kubectl apply -f KubeArmorHostPolicy.yaml
 	cd $(CURDIR); sudo rm -f /tmp/kubearmor.log
-	cd $(CURDIR); sudo -E ./kubearmor -logPath=/tmp/kubearmor.log -enableKubeArmorPolicy=false -enableKubeArmorHostPolicy=true
+	cd $(CURDIR); sudo -E ./kubearmor -logPath=/tmp/kubearmor.log -enableKubeArmorPolicy=false -enableKubeArmorHostPolicy=true -k8s=false
```
Comment (nyrahul) [inline KubeArmor/Makefile]: Cannot set k8s to false in the Makefile ... This will impact regular execution.

File: KubeArmor/config/config.go [hunk 5]
```diff
@@ -144,9 +150,11 @@ func LoadConfig() error {
 		GlobalCfg.Policy = false
 		GlobalCfg.HostPolicy = true
 	}
+	GlobalCfg.K8sEnv = viper.GetBool(ConfigK8sEnv)
+	kl.IsK8SEnv = viper.GetBool(ConfigK8sEnv)
```
Comment (nyrahul) [inline KubeArmor/config/config.go]: Please store the IsK8sEnv in GlobalCfg and then use GlobalCfg in kl module.

File: KubeArmor/config/config.go [hunk 6]
```diff
@@ -28,6 +28,7 @@ type KubearmorConfig struct {
 	Policy     bool // Enable/Disable policy enforcement
 	HostPolicy bool // Enable/Disable host policy enforcement
 	KVMAgent   bool // Enable/Disable KVM Agent
+	K8sEnv     bool // Is k8s env ?
```
Comment (nyrahul) [inline KubeArmor/config/config.go:30]: This variable is not used currently. Please set this variable.

File: KubeArmor/common/common.go [hunk 7]
```diff
@@ -23,6 +23,9 @@ import (
 // == Common == //
 // ============ //
 
+// IsK8SEnv used for specifying k8s environment
+var IsK8SEnv bool
+
```
Comment (nyrahul) [inline KubeArmor/common/common.go]:
```suggestion
```
Remove this var and use GlobalCfg.K8sEnv bool in the check below.

File: KubeArmor/common/common.go [hunk 8]
```diff
@@ -329,9 +332,15 @@ func GetExternalIPAddr() string {
 
 // IsK8sLocal Function
 func IsK8sLocal() bool {
+	if !IsK8SEnv {
```
Comment (nyrahul) [inline KubeArmor/common/common.go]:
```suggestion
	if !cfg.K8sEnv {
```

**Review by nyrahul — CHANGES_REQUESTED** [review]

File: KubeArmor/common/common.go (hunk 7 above)
Comment (seswarrajan) [reply KubeArmor/common/common.go]: It is not possible to include config in common as it causes circular dependency. Hence this global variable.

File: KubeArmor/common/common.go (hunk 8 above)
Comment (seswarrajan) [reply KubeArmor/common/common.go]: It is not possible to include config in common as it causes circular dependency. Hence this global variable.

File: KubeArmor/config/config.go (hunk 5 above)
Comment (seswarrajan) [reply KubeArmor/config/config.go]: It is not possible to include config in common as it causes circular dependency. Hence this global variable.

File: KubeArmor/Makefile (hunk 4 above)
Comment (seswarrajan) [reply KubeArmor/Makefile]: But this is got Kubearmor run-host-only. When run host-only is invoked via make and also .kube/config is present, then host mode will not work.

**Review by nyrahul — APPROVED** [review]

**Pull request merged**